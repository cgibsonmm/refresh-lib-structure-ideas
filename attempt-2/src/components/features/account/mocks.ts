import { rest } from 'msw';
import { mockResolvers } from '@/utils/TestUtils';
import { fetchAccountMock } from './account.mock';
import {
  RESET_PASSWORD_ENDPOINT,
  REQUEST_RESET_PASSWORD_EMAIL_ENDPOINT,
} from './constants';
import { FETCH_ACCOUNT_ENDPOINT } from './useAccount';
import { UPDATE_SETTING_ENDPOINT } from './useUserSetting';
import type { RestRequest, RestContext, ResponseResolver } from 'msw';

const requestResetPasswordEmailResponses = {
  success: {
    account: {
      message:
        'Great. We will send you instructions if an account with that email exists.',
    },
    meta: { status: 200 },
  },
  badRequest: {
    account: { errors: ['Missing Parameters'] },
    meta: { status: 400 },
  },
};

const fetchAccountResponses = {
  success: fetchAccountMock,
  badRequest: {
    account: { errors: ['Missing Parameters'] },
    meta: { status: 400 },
  },
};

const updateSettingResponses = {
  success: {
    message: 'Ok',
    meta: { status: 200 },
  },
  badRequest: {
    message: 'Bad Request',
    meta: { status: 400 },
  },
};

const resetPasswordResponses = {
  success: {},
  badRequest: {
    account: {
      messages: [
        'The reset token has expired. Please try to reset your password again.',
      ],
    },
    meta: { status: 400 },
  },
};

const sendCancellationRequestResponses = {
  success: {},
  badRequest: {
    account: {
      messages: ['Could not cancel account'],
    },
  },
};

const requestResetPasswordEmailResolvers = mockResolvers(
  requestResetPasswordEmailResponses,
  {
    default: async (req, res, ctx) => {
      const body = await req.arrayBuffer();
      const parsedBody = new URLSearchParams(body.toString());

      const isInvalidContentType =
        req.headers.get('Content-Type') !== 'application/x-www-form-urlencoded';
      const isMissingBody = !parsedBody;
      // Because of how we're parsing the request body here, undefined email values were being converted to the string "undefined," and failing to trigger the error response. Adding an extra check for the literal string "undefined" solves the issue, but this may be an indication that we should be parsing 'x-www-form-urlencoded' requests differently in our mocks. - c.j.
      const isMissingEmail =
        parsedBody.get('account[email]') === (undefined || 'undefined');
      const isInvalidRequest =
        isInvalidContentType || isMissingBody || isMissingEmail;

      if (isInvalidRequest)
        return res(
          ctx.status(400),
          ctx.json(requestResetPasswordEmailResponses.badRequest)
        );
      else
        return res(
          ctx.status(200),
          ctx.json(requestResetPasswordEmailResponses.success)
        );
    },
  }
);

const resetPasswordResolvers = mockResolvers(resetPasswordResponses, {
  default: async (req, res, ctx) => {
    const validPassword = 'VALID_TEST_PASSWORD';
    const validToken = 'VALID_TEST_TOKEN';
    const body = await req.arrayBuffer();
    const parsedBody = new URLSearchParams(body.toString());
    const isInvalidContentType =
      req.headers.get('Content-Type') !== 'application/x-www-form-urlencoded';
    const hasInvalidParams =
      parsedBody.get('account[password]') !== validPassword ||
      parsedBody.get('account[password_confirmation]') !== validPassword ||
      parsedBody.get('password_reset_token') !== validToken;
    const isInvalidRequest = isInvalidContentType || hasInvalidParams;
    if (isInvalidRequest)
      return res(ctx.status(400), ctx.json(resetPasswordResponses.badRequest));
    else return res(ctx.status(200), ctx.json(resetPasswordResponses.success));
  },
});

const sendCancellationRequestResolvers = mockResolvers(
  sendCancellationRequestResponses,
  {
    default: async (req, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json(sendCancellationRequestResponses.success)
      );
    },
  }
);

const fetchAccountResolvers = mockResolvers(fetchAccountResponses);

type MockResolver = ResponseResolver<RestRequest<URLSearchParams>, RestContext>;

function requestResetPasswordEmailHandler(resolver: MockResolver) {
  return rest.post(REQUEST_RESET_PASSWORD_EMAIL_ENDPOINT, resolver);
}

function resetPasswordHandler(resolver: MockResolver) {
  return rest.post(RESET_PASSWORD_ENDPOINT, resolver);
}

function fetchAccountHandler(resolver: MockResolver) {
  const fetchAccountUrl = FETCH_ACCOUNT_ENDPOINT;
  return rest.get(fetchAccountUrl, resolver);
}

const updateSettingResolvers = mockResolvers(updateSettingResponses);

function updateSettingHandler(resolver: MockResolver) {
  const updateSettingUrl = UPDATE_SETTING_ENDPOINT;
  return rest.patch(updateSettingUrl, resolver);
}

function sendCancellationRequestHandler(resolver: MockResolver) {
  const sendCancellationRequestUrl = '/api/v5/account/cancel';
  return rest.post(sendCancellationRequestUrl, resolver);
}

export const handlers = {
  requestResetPasswordEmail: requestResetPasswordEmailHandler,
  resetPassword: resetPasswordHandler,
  fetchAccount: fetchAccountHandler,
  updateSetting: updateSettingHandler,
  sendCancellationRequest: sendCancellationRequestHandler,
};

export const resolvers = {
  requestResetPasswordEmail: requestResetPasswordEmailResolvers,
  resetPassword: resetPasswordResolvers,
  fetchAccount: fetchAccountResolvers,
  updateSetting: updateSettingResolvers,
  sendCancellationRequest: sendCancellationRequestResolvers,
};

export const responses = {
  requestResetPasswordEmail: requestResetPasswordEmailResponses,
  resetPassword: resetPasswordResponses,
  fetchAccount: fetchAccountResponses,
  updateSetting: updateSettingResponses,
  sendCancellationRequest: sendCancellationRequestResponses,
};
