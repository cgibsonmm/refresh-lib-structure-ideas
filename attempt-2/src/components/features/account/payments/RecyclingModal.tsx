import { useContext, useEffect, useState } from 'react';
import { useSession } from '@/auth/index';
import { AppConstants } from '@/context/AppConstants';
import { Button, Modal, Stack, Text } from '@/theme/index';

interface RecyclingModalProps {
  /** If true, the modal is open. */
  open: boolean;
  /**
   * Callback fired when the user tries to close the modal by either
   * clicking on the close button, on the backdrop, on the **Not now** button,
   * or by pressing the escape key.
   */
  onClose: () => void;
  /** Callback fired when the user clicks on the **Update payment info now** button. */
  openUpdatePayment: () => void;
}

export const RecyclingModal = ({
  open,
  onClose,
  openUpdatePayment,
}: RecyclingModalProps) => {
  const [recyclingReportsRemaining, setRecyclingReportsRemaining] =
    useState(false);
  const {
    session: { account, userHasRecyclingReportsRemaining },
  } = useSession();
  useEffect(() => {
    setRecyclingReportsRemaining(!!userHasRecyclingReportsRemaining);
  }, [account]);

  const { paymentRequired, brandName } = useContext(AppConstants);
  return (
    <Modal hasCloseIcon open={open} onClose={onClose}>
      <Stack
        direction="column"
        alignItems="center"
        textAlign="center"
        gap={1.25}
        maxWidth={300}
      >
        <img src={paymentRequired} alt="Payment Required" />
        <Text variant="h2">Payment Update Required</Text>
        <Text variant="subtitle2">
          {recyclingReportsRemaining
            ? `We were unable to bill your account, but we have left it active as a courtesy. Please update your payment method on file to continue accessing ${brandName}`
            : `We were unable to bill your account. Please update your payment method on file to continue accessing ${brandName}`}
        </Text>
        <Button variant="contained" color="primary" onClick={openUpdatePayment}>
          Update payment info now
        </Button>
        <Text
          variant="h6"
          color="primary"
          sx={{ cursor: 'pointer' }}
          onClick={onClose}
        >
          Not now
        </Text>
      </Stack>
    </Modal>
  );
};
