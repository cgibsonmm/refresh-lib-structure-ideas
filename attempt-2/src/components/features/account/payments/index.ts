export { RecyclingModal } from './RecyclingModal';
export { ViewOrderHistory } from './ViewOrderHistory';
export { PayPalRedirectionModal } from './paymentsUpdateModal/PaymentsComponents/PayPalRedirectionModal';
export { RadioPaymentsModal } from './paymentsUpdateModal/PaymentsComponents/RadioPaymentsModal';
export {
  useProcessCard,
  useProcessPayPal,
  useGetPlans,
} from './useSubscriptionUpgrade';
