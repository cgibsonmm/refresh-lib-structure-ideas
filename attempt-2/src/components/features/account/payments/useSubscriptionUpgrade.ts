import { useContext } from 'react';
import { useMutation, useQuery, useQueryClient } from 'react-query';
import { AppConfig } from '@/context/AppConfig';
import { request } from '@/utils/requestHelpers';

const UPGRADE_OPTIONS_URL = '/api/v5/subscription/upgrade_options.json';
const UPGRADE_URL = '/api/v5/payments/credit_card_payment.json';
const UPGRADE_URL_PAYPAL = '/api/v5/payments/paypal_payment.json';
const REACTIVATE_URL = '/api/v5/subscription/reactivate.json';
const DOWNSELL_PLAN_SUBSCRIPTIONS_URL =
  '/api/v5/subscription/downsell_options.json';

export function useGetPlans() {
  return useQuery(['getPlans'], () => {
    const upgradeOptionsResponse = request(UPGRADE_OPTIONS_URL, {
      headers: new Headers({
        'content-type': 'application/x-www-form-urlencoded',
      }),
      method: 'GET',
    });
    return upgradeOptionsResponse;
  });
}

export function useGetDownsellPlan() {
  return useQuery(['getDownsellPlan'], () => {
    const downsellPlanResponse = request(DOWNSELL_PLAN_SUBSCRIPTIONS_URL, {
      headers: new Headers({
        'content-type': 'application/x-www-form-urlencoded',
      }),
      method: 'GET',
    });
    return downsellPlanResponse;
  });
}

export const useProcessCard = () => {
  const queryClient = useQueryClient();
  const { logError } = useContext(AppConfig);

  return useMutation(
    ({
      creditCardId,
      plan,
    }: {
      creditCardId: number;
      plan: { unique_key: string };
    }) => {
      const formData = new URLSearchParams();
      formData.set('cart[products][SubscriptionPlan][]', plan.unique_key);
      formData.set('credit_card_id', String(creditCardId));

      return request(
        UPGRADE_URL,
        {
          method: 'POST',
          body: formData,
        },
        undefined,
        _paymentErrorHandler
      );
    },
    {
      onSuccess: (data) => {
        if (data?.meta?.status) {
          queryClient.invalidateQueries({ queryKey: ['account'] });
          queryClient.invalidateQueries({ queryKey: ['getPaymentMethod'] });
        }
      },
      onError: (error: Error) => {
        logError('Error processing credit card payment', error);
      },
    }
  );
};

export const useProcessPayPal = () => {
  const queryClient = useQueryClient();
  const { logError } = useContext(AppConfig);

  return useMutation(
    ({
      paymentAgreementId,
      plan,
    }: {
      paymentAgreementId: number;
      plan: { unique_key: string };
    }) => {
      const formData = new URLSearchParams();
      formData.set('cart[products][SubscriptionPlan][]', plan.unique_key);
      formData.set('paypal[id]', String(paymentAgreementId));

      return request(
        UPGRADE_URL_PAYPAL,
        {
          method: 'POST',
          body: formData,
        },
        undefined,
        _paymentErrorHandler
      );
    },
    {
      onSuccess: (data) => {
        if (data?.meta?.status) {
          queryClient.invalidateQueries({ queryKey: ['account'] });
          queryClient.invalidateQueries({ queryKey: ['getPaymentMethod'] });
        }
      },
      onError: (error: Error) => {
        logError('Error processing PayPal payment', error);
      },
    }
  );
};

export const useReactivateAccountTimeLeft = () => {
  return useMutation((confirm: string) => {
    const formData = new URLSearchParams();
    formData.set('confirm', confirm);
    return request(REACTIVATE_URL, {
      method: 'POST',
      body: formData,
    });
  });
};

const _paymentErrorHandler = async (response: Response): Promise<Error> => {
  const data = await response.json();

  const errors = data.checkout_response?.errors || data.errors || [];

  return Promise.reject(
    new Error(
      `Error processing payment (${response.status} - ${response.statusText})`,
      { cause: errors }
    )
  );
};
