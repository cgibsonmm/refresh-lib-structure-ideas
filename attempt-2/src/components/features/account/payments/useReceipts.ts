import { useQuery } from 'react-query';
import { Receipt } from '@/account/Interfaces';
import { request } from '@/utils/requestHelpers';

//Hook to get all receipts.
export function useReceipts() {
  return useQuery(['getReceipts'], async () => {
    const paymentQuery = await request('/api/v5/receipts', {
      headers: new Headers({
        'content-type': 'application/x-www-form-urlencoded',
      }),
      method: 'GET',
    });
    formatReceipt(paymentQuery);
    return paymentQuery;
  });
}

//Format date, method and amount.
const formatReceipt = ({ receipts }: { receipts: Receipt[] }) => {
  receipts.forEach((receipt: Receipt) => {
    formatDate(receipt);
    formatAmount(receipt);
    formatMethod(receipt);
  });
};

const formatDate = (receipt: Receipt) => {
  const date = new Date(receipt.created_at);
  receipt.created_at = date.toLocaleDateString();
};

const formatAmount = (receipt: Receipt) => {
  const amountUSD = Number(receipt.amount) / 100;
  const usdFormater = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
  });
  receipt.amount = usdFormater.format(amountUSD);
};

const formatMethod = (receipt: Receipt) => {
  const method = receipt.method;
  if (method === 'BRAINTREE PAYMENT') {
    receipt.method = 'Apm'; //TODO: Handle APMs
    return;
  }
  const methodPart = receipt.method.split(' ');
  if (methodPart[1]) {
    receipt.method = methodPart[1].slice(1);
  }
};
