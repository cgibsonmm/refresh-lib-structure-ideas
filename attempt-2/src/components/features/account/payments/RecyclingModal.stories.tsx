import { Meta, StoryObj } from '@storybook/react';
import { RecyclingModal } from './RecyclingModal';

const meta: Meta<typeof RecyclingModal> = {
  title: 'Features/Payments/RecyclingModal',
  component: RecyclingModal,
  parameters: {
    docs: {
      story: {
        inline: false,
        iframeHeight: 400,
      },
    },
  },
  argTypes: {
    openUpdatePayment: { action: 'openUpdatePayment' },
    onClose: { action: 'onClose' },
  },
  args: {
    open: true,
  },
};

export default meta;

type Story = StoryObj<typeof RecyclingModal>;

export const Default: Story = {};
