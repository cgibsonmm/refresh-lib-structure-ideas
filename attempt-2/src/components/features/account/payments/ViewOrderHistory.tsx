import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';
import { styled } from '@mui/system';
import { Receipt } from '@/account/Interfaces';
import { useReceipts } from '@/account/payments/useReceipts';
import { Text, Table, Accordion, Icons } from '@/theme/index';

export const FlexDiv = styled('div')({
  display: 'flex',
  alignItems: 'center',
});

export const ValueDiv = styled('div')({
  width: '50%',
  textAlign: 'right',
  display: 'flex',
  justifyContent: 'flex-end',
  alignItems: 'center',
});

const renderHeader = (isMobile: boolean) => {
  //Mobile table headers
  if (isMobile) {
    return ['TRANSACTION'];
  }

  //Desktop table headers.
  return ['DATE', 'TRANSACTION', 'BILLED', 'AMOUNT'];
};

const PaymentMethod = (props: Receipt) => {
  return (
    <FlexDiv>
      <Icons.CreditCard />
      <Text variant="caption">{props?.method}</Text>
    </FlexDiv>
  );
};

const renderContent = (object: Receipt[], isMobile: boolean) => {
  if (object === undefined || object.length === 0) {
    //No receipts.
    return [[]];
  }

  //Mobile table content
  if (isMobile) {
    return object.map((value: Receipt) => {
      return [
        <>
          <Text variant="h3">{value.created_at}</Text>
          <Text variant="body2">{value?.items[0]?.description}</Text>
          <FlexDiv>
            <div style={{ width: '50%' }}>{PaymentMethod(value)}</div>
            <ValueDiv>
              <Text variant="caption">{value.amount}</Text>
            </ValueDiv>
          </FlexDiv>
        </>,
      ];
    });
  }

  //Desktop table content
  return object.map((value: Receipt) => {
    return [
      value?.created_at,
      value?.items[0]?.description,
      PaymentMethod(value),
      <Text marginLeft={1}>{value?.amount}</Text>,
    ];
  });
};

/**
 * This component is used to display the user's order history which includes the date, transaction,
 * payment method used, and amount.
 */
export const ViewOrderHistory = () => {
  const { data, isError } = useReceipts();
  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down('sm'));

  return (
    <Accordion display="View order history" data-cy="order_history">
      <Table
        data-cy="order_history_data"
        headers={
          data?.receipts?.length === 0 || data === undefined || isError
            ? ['No payment history found']
            : renderHeader(isMobile)
        }
        rows={renderContent(data?.receipts, isMobile)}
      />
    </Accordion>
  );
};
