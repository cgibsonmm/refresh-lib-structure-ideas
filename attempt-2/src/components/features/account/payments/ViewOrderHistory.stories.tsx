import { StoryObj, Meta } from '@storybook/react';
import { rest } from 'msw';
import { fetchAccountMock } from '@/account/account.mock';
import { ViewOrderHistory } from '@/account/payments/ViewOrderHistory';
import { receiptsResponse } from '@/mocks/receipts';

const meta: Meta<typeof ViewOrderHistory> = {
  title: 'Features/Payments/ViewOrderHistory',
  component: ViewOrderHistory,
  parameters: {
    sessionProvider: true,
    msw: {
      handlers: [
        rest.get('/api/v5/receipts', (req, res, ctx) => {
          return res(ctx.json(receiptsResponse));
        }),
        rest.get('/api/v5/account', (req, res, ctx) => {
          return res(ctx.status(200), ctx.json(fetchAccountMock));
        }),
      ],
    },
  },
};

export default meta;

type Story = StoryObj<typeof ViewOrderHistory>;

export const Default: Story = {};
