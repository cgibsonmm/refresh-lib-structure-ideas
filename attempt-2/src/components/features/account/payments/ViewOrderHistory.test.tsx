import { render, screen } from '@testing-library/react';
import { ViewOrderHistory } from '@/account/payments/ViewOrderHistory';
import { useReceipts } from '@/account/payments/useReceipts';
import { receiptsResponse } from '@/mocks/receipts';
import { disableReactQueryErrorLogs, withProviders } from '@/utils/TestUtils';
disableReactQueryErrorLogs();

jest.mock('./useReceipts.ts');

const useReceiptsMock = useReceipts as jest.MockedFunction<typeof useReceipts>;

test(`It renders the correct headers title when there's no error and there is data`, () => {
  (useReceiptsMock as jest.Mock).mockReturnValue({
    isError: false,
    data: receiptsResponse,
  });

  const { getByText } = render(withProviders(ViewOrderHistory));

  expect(getByText('TRANSACTION')).toBeInTheDocument();
});

test('It renders "No payment history found" when data is undefined', () => {
  (useReceiptsMock as jest.Mock).mockReturnValue({
    isError: false,
    data: undefined,
  });

  const { getByText } = render(withProviders(ViewOrderHistory));

  expect(getByText('No payment history found')).toBeInTheDocument();
});

test('It renders "No payment history found" when error is true', () => {
  (useReceiptsMock as jest.Mock).mockReturnValue({
    isError: true,
    data: receiptsResponse,
  });

  const { getByText } = render(withProviders(ViewOrderHistory));

  expect(getByText('No payment history found')).toBeInTheDocument();
});

test('The PaymentSelection screen renders on the document', () => {
  (useReceiptsMock as jest.Mock).mockReturnValue({
    isError: false,
    data: receiptsResponse,
  });

  const { asFragment } = render(withProviders(ViewOrderHistory));
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText('TRANSACTION')).toBeInTheDocument();
});
