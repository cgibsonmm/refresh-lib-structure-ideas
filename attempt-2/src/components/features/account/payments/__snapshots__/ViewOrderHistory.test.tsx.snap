// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The PaymentSelection screen renders on the document 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-kjs9sy-MuiPaper-root-MuiAccordion-root"
    data-cy="order_history"
  >
    <div
      aria-controls="panel1a-content"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
      id="panel1a-header"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
        >
          View order history
        </p>
      </div>
      <div
        class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-1bptiuh-MuiSvgIcon-root"
          data-testid="ExpandCircleDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 13.5L7.5 11l1.42-1.41L12 12.67l3.08-3.08L16.5 11 12 15.5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel1a-header"
            class="MuiAccordion-region"
            id="panel1a-content"
            role="region"
          >
            <div
              class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              <table
                class="MuiTable-root css-6zcppr-MuiTable-root"
                data-cy="order_history_data"
                headers="DATE,TRANSACTION,BILLED,AMOUNT"
              >
                <thead
                  class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      DATE
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      TRANSACTION
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      BILLED
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                      scope="col"
                    >
                      AMOUNT
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      2023-01-24T09:46:59.000-05:00
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      Unlimited Reports - Free (with a 1 week trial period)
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      <div
                        class="css-70qvj9"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="CreditCardIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 4H4c-1.11 0-1.99.89-1.99 2L2 18c0 1.11.89 2 2 2h16c1.11 0 2-.89 2-2V6c0-1.11-.89-2-2-2zm0 14H4v-6h16v6zm0-10H4V6h16v2z"
                          />
                        </svg>
                        <span
                          class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
                        >
                          VISA x-7872
                        </span>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                      >
                        7023
                      </p>
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      2023-01-20T09:46:59.000-05:00
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      Unlimited Reports - Free (with a 1 week trial period)
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      <div
                        class="css-70qvj9"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="CreditCardIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 4H4c-1.11 0-1.99.89-1.99 2L2 18c0 1.11.89 2 2 2h16c1.11 0 2-.89 2-2V6c0-1.11-.89-2-2-2zm0 14H4v-6h16v6zm0-10H4V6h16v2z"
                          />
                        </svg>
                        <span
                          class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
                        >
                          VISA x-7872
                        </span>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-hgini7-MuiTableCell-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                      >
                        2973
                      </p>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
